src/mpyl/utilities/s3/__init__.py

```diff
@@ -0,0 +1,82 @@
+"""A client to connect to an S3 bucket"""
+
+import pathlib
+from typing import Dict
+from logging import Logger
+from os import walk
+import boto3
+from botocore.exceptions import ClientError
+
+
+class S3Client:
+
+    def __init__(self, logger: Logger, config: Dict, bucket_name: str, root_path: str):
+        """
+        Creates a client that provides a wrapper for uploading a directory to an S3 bucket
+
+        :param config: the mpyl_config containing an AWS access key ID and secret key
+        :param bucket_name: the name of the bucket this client will operate on
+        :param root_path: the root path on the bucket in which files will be uploaded to
+        """
+        s3_config = config.get('s3', None)
+        if not s3_config or not s3_config.get('accessKeyId') or not s3_config.get('secretAccessKey'):
+            raise ValueError('S3 configuration not set')
+
+        self._client = boto3.client(service_name='s3',
+                                    region_name='eu-central-1',
+                                    aws_access_key_id=s3_config['accessKeyId'],
+                                    aws_secret_access_key=s3_config['secretAccessKey'])
+        self._bucket_name = bucket_name
+        self._root_path = root_path
+        self._logger = logger
+
+    def upload_directory(self, directory: str):
+        """
+        Uploads all files in a directory and its subdirectories to the S3 bucket
+
+        Note: boto3 does not provide a sync method, so we have to walk the directory and upload each file individually
```

Comment:
The lack of a sync is one of the main downsides of Amazons python SDK. There are other libraries that implement it, but it seems like a good idea to stick with the official SDK and it's not too big of a deal to get a similar result.

The only downside with this implementation is, with subsequent deployments to a PR environment, if an asset was deleted it will still remain in the bucket. I don't think this is a common enough occurrence to worry about and it only effects the test bucket (in Vandebron's case)